### DISCUSSION TITLE
Issue with some letters not joining in arabic script

### DISCUSSION THREAD

**Discussion by Madour** [harfbuzz/harfbuzz#4177] [Q&A] [answered]
I am trying to implement arabic text rendering with opengl using freetyp2 and harfbuzz.

I used this code to get me started : https://github.com/harfbuzz/harfbuzz-tutorial

So far, I am getting this result:

As you can see, there are 2 letters that are detached but should be attached to the left.
![Capture d'écran 2023-03-18 080306](https://user-images.githubusercontent.com/11854124/226090677-cc771cef-77cb-4415-be37-8d02c4e37010.png)

I printed the name, position (pos + offset), advance and texture width of each glyph after shaping:
```
sukun         uni0652 @  130.86 ( 100.00 + 31), +0.00   (w= 48)
meem          uniFEE2 @  100.00 ( 100.00 +  0), +165.44 (w=151)
damma         uni064F @  269.34 ( 265.44 +  4), +0.00   (w= 76)
kaaf          uniFEDC @  265.44 ( 265.44 +  0), +150.39 (w=165)
yaa'          uniFEF4 @  415.83 ( 415.83 +  0), +88.09  (w=103)
fatha         uni064E @  501.19 ( 503.92 -  3), +0.00   (w= 81)
laam          uniFEE0 @  503.92 ( 503.92 +  0), +83.98  (w= 98)
fatha         uni064E @  601.97 ( 587.91 + 14), +0.00   (w= 81)
3ain          uniFECB @  587.91 ( 587.91 +  0), +143.17 (w=148)
                space @  731.08 ( 731.08 +  0), +78.12  (w=  0)
damma         uni064F @  837.72 ( 809.20 + 29), +0.00   (w= 76)
meem          uni0645 @  809.20 ( 809.20 +  0), +139.06 (w=107)
laam+alif     uniFEFC @  918.97 ( 948.27 - 29), +136.53 (w=153)
shadda+fatha  uniFC60 @ 1143.98 (1084.80 + 59), +0.00   (w= 67)
seen          uniFEB4 @ 1084.80 (1084.80 +  0), +200.00 (w=214)
laam          uniFEDF @ 1284.80 (1284.80 +  0), +75.00  (w= 77)
alif          uni0627 @ 1359.80 (1359.80 +  0), +62.11  (w= 27)
```

I noticed, for the last meem, the advance value is bigger than the texture width, which could explain the issue, but why ?
And for the detached seen, it seems to be caused by the offset on the laam+alif glyph.

Here is the code that creates the vertices; as I mentioned, I based my code on the harfbuzz-tutorial code, and the loop is similar:
```c++
double current_x = 100;
double current_y = 300;
for (unsigned int i = 0; i < len; i++)
{
    hb_codepoint_t gid = info[i].codepoint;
    auto& this_pos = pos[i];

    char glyph_name[32];
    hb_font_get_glyph_name(hb_font, gid, glyph_name, 32);

    auto& glyph = font.getGlyph(gid, FONT_SIZE); // this comes from my code that creates the font texture

    double pos_x = current_x + this_pos.x_offset / 64.;
    double pos_y = current_y - this_pos.y_offset / 64. - glyph.base; // glyph.base is horiBearingY from freetype metrics
    double adv_x = this_pos.x_advance / 64.;
    double adv_y = this_pos.y_advance / 64.;

    double tex_x = glyph.tex_pos.x;
    double tex_y = glyph.tex_pos.y;
    double tex_w = glyph.tex_size.x;
    double tex_h = glyph.tex_size.y;

    text_va.pushQuad({{
        {{pos_x,         pos_y},
         {tex_x,         tex_y}},
        
        {{pos_x + tex_w, pos_y},
         {tex_x + tex_w, tex_y}},
        
        {{pos_x + tex_w, pos_y + tex_h},
         {tex_x + tex_w, tex_y + tex_h}},
        
        {{pos_x,         pos_y + tex_h},
         {tex_x,         tex_y + tex_h}}
    }});

    auto off = this_pos.x_offset / 64.;
    printf("%8s @ %4.2f (%4.2f %c %2.f), %+3.2f (w=%3.f)\n", glyph_name, pos_x, current_x, off < 0 ? '-' : '+', abs(off), adv_x, tex_w);

    current_x += adv_x;
    current_y += adv_y;
}
```

Here is the font texture generated and used to render the text, it looks fine:
![image](https://user-images.githubusercontent.com/11854124/226091297-e63cde4c-47e8-4840-88fa-0244ab126e06.png)

Any help is welcome :).

**Comment by khaledhosny** [comment, marked as answer]
What is the font you are using? Can you reproduce the issue with `hb-view` or `hb-shape`?

**Reply by khaledhosny** [reply]
I think you are probably not accounting for the side bearing of the glyphs, your vertices use the glyph bounding box width. I think you need to handle `horiBearingX` in the x position like you do with `horiBearingY`.

**Reply by Madour** [reply]
Thank you for the quick response!

> What is the font you are using ?

I am using AdobeArabic-Regular.otf

> Can you reproduce the issue with hb-view or hb-shape ?

I tried using hb-view to generate a png, but all I get is an image with "???". It may be related to encoding ? I even have to encode my cpp file in UTF-16, because otherwise UTF-8 does not work, even though I am using wide strings.

> I think you are probably not accounting for the side bearing of the glyphs, your vertices use the glyph bounding box width. I think you need to handle horiBearingX in the x position like you do with horiBearingY.

This totally fixed the issue ! I don't know why I didn't think of adding the X bearing ^^'.

But now I have another question: why do I still have to get and add metrics from freetype glyphs? 
Isn't it harfbuzz role to provide the final glyphs position after shaping? Why isn't the bearing included in harfbuzz `pos.x_offset`  and `pos.y_offset` for example?

**Reply by khaledhosny** [reply]
HarfBuzz positions are relative to the glyph origin, you need to add the side bearing to your drawing code because your atlas generation code strips them and gives you an atlas without any side bearings. This is very specific to the way you render your text and HarfBuzz shaping code wouldn’t know anything about it.
